qiime/plugin/plugin.py

```diff
@@ -59,23 +65,34 @@ def actions(self):
         actions.update(self.visualizers)
         return types.MappingProxyType(actions)
 
-    def register_data_layout(self, data_layout):
-        if not isinstance(data_layout, DataLayout):
-            raise TypeError(
-                "`data_layout` must be an instance of type %r, not type %r."
-                % (DataLayout.__name__, type(data_layout).__name__))
-        # Finalize data layout to prevent additional files from being
-        # registered on the data layout after it has been registered.
-        data_layout.finalize()
-        name = data_layout.name
-        version = data_layout.version
-        self.data_layouts[(name, version)] = data_layout
-
-    def register_data_layout_reader(self, name, version, view_type, reader):
-        self.data_layout_readers[(name, version, view_type)] = reader
-
-    def register_data_layout_writer(self, name, version, view_type, writer):
-        self.data_layout_writers[(name, version, view_type)] = writer
+    def register_transformer(self, _fn=None, *, restrict=None):
+        """
+        A transformer has the type Callable[[type], type]
+        """
+        def decorator(transformer):
+            annotations = transformer.__annotations__.copy()
+            if len(annotations) != 2:
+                raise TypeError()
+            if type(annotations['return']) is tuple:
+                raise TypeError()
```

Comment:
two types may be provided without a return. @jairideout has a utility helper somewhere